File: freightmas/freightmas/report/stock_balance_freightmas/stock_balance_freightmas.js
// Copyright (c) 2025, Zvomaita Technologies (Pvt) Ltd and contributors
// For license information, please see license.txt

frappe.query_reports["Stock Balance FreightMas"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "date_range",
            "label": __("Date Range"),
            "fieldtype": "Select",
            "options": [
                "",
                "Today",
                "Yesterday",
                "This Week",
                "Last Week",
                "This Month",
                "Last Month",
                "This Year",
                "Last Year",
                "Custom"
            ],
            "default": "This Month",
            "on_change": function () {
                let date_range = frappe.query_report.get_filter_value('date_range');
                let today = frappe.datetime.get_today();
                let from_date, to_date;

                switch (date_range) {
                    case "Today":
                        from_date = to_date = today;
                        break;
                    case "Yesterday":
                        from_date = to_date = frappe.datetime.add_days(today, -1);
                        break;
                    case "This Week":
                        from_date = frappe.datetime.week_start();
                        to_date = frappe.datetime.week_end();
                        break;
                    case "Last Week":
                        from_date = frappe.datetime.add_days(frappe.datetime.week_start(), -7);
                        to_date = frappe.datetime.add_days(frappe.datetime.week_end(), -7);
                        break;
                    case "This Month":
                        from_date = frappe.datetime.month_start();
                        to_date = frappe.datetime.month_end();
                        break;
                    case "Last Month":
                        from_date = frappe.datetime.add_months(frappe.datetime.month_start(), -1);
                        to_date = frappe.datetime.add_days(frappe.datetime.month_start(), -1);
                        break;
                    case "This Year":
                        from_date = frappe.datetime.year_start();
                        to_date = frappe.datetime.year_end();
                        break;
                    case "Last Year":
                        let today_parts = today.split('-');
                        let last_year = parseInt(today_parts[0], 10) - 1;
                        from_date = `${last_year}-01-01`;
                        to_date = `${last_year}-12-31`;
                        break;
                    default:
                        from_date = frappe.query_report.get_filter_value('from_date');
                        to_date = frappe.query_report.get_filter_value('to_date');
                }

                if (date_range && date_range !== "Custom" && date_range !== "") {
                    frappe.query_report.set_filter_value('from_date', from_date);
                    frappe.query_report.set_filter_value('to_date', to_date);
                }
            }
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_start(),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_end(),
            "reqd": 1
        },
        {
            "fieldname": "warehouse",
            "label": __("Warehouse"),
            "fieldtype": "Link",
            "options": "Warehouse",
            "get_query": function() {
                var company = frappe.query_report.get_filter_value('company');
                return {
                    "filters": {
                        "company": company
                    }
                }
            }
        },
        {
            "fieldname": "item_code",
            "label": __("Item"),
            "fieldtype": "Link",
            "options": "Item",
            "get_query": function() {
                return {
                    "filters": {
                        "disabled": 0
                    }
                }
            }
        },
        {
            "fieldname": "item_group",
            "label": __("Item Group"),
            "fieldtype": "Link",
            "options": "Item Group"
        },
        {
            "fieldname": "warehouse_type",
            "label": __("Warehouse Type"),
            "fieldtype": "Link",
            "options": "Warehouse Type"
        },
        {
            "fieldname": "include_zero_stock_items",
            "label": __("Include Zero Stock Items"),
            "fieldtype": "Check",
            "default": 0
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Color negative balance quantities in red, positive stays default color
        if (column.fieldname == "balance_qty" && data && data.balance_qty < 0) {
            value = "<span style='color:red'>" + value + "</span>";
        }
        
        // Color negative opening quantities in red
        if (column.fieldname == "opening_qty" && data && data.opening_qty < 0) {
            value = "<span style='color:red'>" + value + "</span>";
        }
        
        // Color positive in_qty values in green
        if (column.fieldname == "in_qty" && data && data.in_qty > 0) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        
        // Color positive out_qty values in blue
        if (column.fieldname == "out_qty" && data && data.out_qty > 0) {
            value = "<span style='color:blue'>" + value + "</span>";
        }
        
        // Positive opening quantities and zero values remain default color (no formatting needed)
        
        return value;
    },

    "onload": function(report) {
        // Set default company if available
        if (frappe.defaults.get_user_default("Company")) {
            report.set_filter_value("company", frappe.defaults.get_user_default("Company"));
        }

        // Add Export buttons
        report.page.add_inner_button('Export to Excel', function() {
            const filters = report.get_filter_values(true);
            const query = encodeURIComponent(JSON.stringify(filters));
            const url = `/api/method/freightmas.api.export_report_to_excel?report_name=Stock Balance FreightMas&filters=${query}`;
            window.open(url);
        }, 'Export');

        report.page.add_inner_button('Export to PDF', function() {
            const filters = report.get_filter_values(true);
            const query = encodeURIComponent(JSON.stringify(filters));
            const url = `/api/method/freightmas.api.export_report_to_pdf?report_name=Stock Balance FreightMas&filters=${query}`;
            window.open(url);
        }, 'Export');

        // Add a stand-alone "Clear Filters" button
        report.page.add_inner_button('Clear Filters', function() {
            // Clear all filters to defaults
            report.filters.forEach(filter => {
                let default_value = filter.df.default || "";
                if (filter.df.fieldname === 'from_date') {
                    default_value = frappe.datetime.month_start();
                } else if (filter.df.fieldname === 'to_date') {
                    default_value = frappe.datetime.month_end();
                } else if (filter.df.fieldname === 'company') {
                    default_value = frappe.defaults.get_user_default("Company") || "";
                } else if (filter.df.fieldname === 'date_range') {
                    default_value = "This Month";
                } else if (filter.df.fieldname === 'include_zero_stock_items') {
                    default_value = 0;
                }
                report.set_filter_value(filter.df.fieldname, default_value);
            });
            
            // Refresh the report after clearing filters
            setTimeout(function() {
                report.refresh();
            }, 500);
        });
    }
};

File: freightmas/freightmas/report/stock_balance_freightmas/stock_balance_freightmas.py
# Copyright (c) 2025, Zvomaita Technologies (Pvt) Ltd and contributors
# For license information, please see license.txt

# import frappe


import frappe
from frappe import _
from frappe.utils import flt, getdate

def execute(filters=None):
    if not filters:
        filters = {}
    
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": _("Item"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 180
        },
        {
            "label": _("Warehouse"),
            "fieldname": "warehouse",
            "fieldtype": "Link",
            "options": "Warehouse",
            "width": 240
        },
        {
            "label": _("Stock UOM"),
            "fieldname": "stock_uom",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": _("Opening Qty"),
            "fieldname": "opening_qty",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": _("In Qty"),
            "fieldname": "in_qty",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": _("Out Qty"),
            "fieldname": "out_qty",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": _("Balance Qty"),
            "fieldname": "balance_qty",
            "fieldtype": "Float",
            "width": 120
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    stock_ledger_entries = get_stock_ledger_entries(filters, conditions)
    iwb_map = get_item_warehouse_map(stock_ledger_entries, filters)
    
    data = []
    for key, value in iwb_map.items():
        item_code, warehouse = key
        
        # Skip items with zero balance if not requested
        if not filters.get("include_zero_stock_items") and not value.get("balance_qty"):
            continue
            
        data.append({
            "item_code": item_code,
            "warehouse": warehouse,
            "stock_uom": value.get("stock_uom"),
            "opening_qty": flt(value.get("opening_qty"), 3),
            "in_qty": flt(value.get("in_qty"), 3),
            "out_qty": flt(value.get("out_qty"), 3),
            "balance_qty": flt(value.get("balance_qty"), 3)
        })
    
    return data

def get_conditions(filters):
    conditions = ""
    
    if filters.get("company"):
        conditions += " AND sle.company = %(company)s"
    if filters.get("warehouse"):
        conditions += " AND sle.warehouse = %(warehouse)s"
    if filters.get("item_code"):
        conditions += " AND sle.item_code = %(item_code)s"
    if filters.get("item_group"):
        conditions += " AND i.item_group = %(item_group)s"
    if filters.get("warehouse_type"):
        conditions += " AND w.warehouse_type = %(warehouse_type)s"
    
    return conditions

def get_stock_ledger_entries(filters, conditions):
    return frappe.db.sql("""
        SELECT
            sle.item_code,
            sle.warehouse,
            sle.posting_date,
            sle.actual_qty,
            i.stock_uom
        FROM
            `tabStock Ledger Entry` sle
        INNER JOIN
            `tabItem` i ON sle.item_code = i.name
        INNER JOIN
            `tabWarehouse` w ON sle.warehouse = w.name
        WHERE
            sle.docstatus < 2
            AND sle.posting_date <= %(to_date)s
            {conditions}
        ORDER BY
            sle.item_code, sle.warehouse, sle.posting_date
    """.format(conditions=conditions), filters, as_dict=1)

def get_item_warehouse_map(sle, filters):
    iwb_map = {}
    from_date = getdate(filters.get("from_date"))
    to_date = getdate(filters.get("to_date"))
    
    for d in sle:
        key = (d.item_code, d.warehouse)
        if key not in iwb_map:
            iwb_map[key] = {
                "opening_qty": 0.0,
                "in_qty": 0.0,
                "out_qty": 0.0,
                "balance_qty": 0.0,
                "stock_uom": d.stock_uom
            }
        
        qty_dict = iwb_map[key]
        
        if d.posting_date < from_date:
            qty_dict["opening_qty"] += flt(d.actual_qty)
        elif d.posting_date >= from_date and d.posting_date <= to_date:
            if flt(d.actual_qty) > 0:
                qty_dict["in_qty"] += flt(d.actual_qty)
            else:
                qty_dict["out_qty"] += abs(flt(d.actual_qty))
        
        qty_dict["balance_qty"] += flt(d.actual_qty)
    
    return iwb_map